### PR TITLE
tipi.build support for tomlplusplus

### PR DESCRIPTION
**Pull request by pysco68**
Hello from [tipi.build](https://tipi.build) :D,

I just added `tipi` support to your project and was wondering if you'd like to merge so everyone gets a new way of using your cool project.

If there's any issue with this I'll be happy to sort it out

Best from Zürich,
Yannic / pysco68

**Pre-merge checklist**

- [x] I've read [CONTRIBUTING.md]
- [x] I've rebased my changes against the current HEAD of `origin/master` (if necessary)
- [x] I've added new test cases to verify my change (added a CI job which uses the tipi-ubuntu docker container)
- [ ] ~~I've regenerated toml.hpp~~
- [x] I've updated any affected documentation
- [x] I've rebuilt and run the tests with at least one of:
    - [x] Clang 6 or higher
    - [ ] GCC 7 or higher
    - [x] MSVC 19.20 (Visual Studio 2019) or higher
- [ ] ~~I've added my name to the list of contributors in [README.md](https://github.com/marzer/tomlplusplus/blob/master/README.md)~~

[CONTRIBUTING.md]: https://github.com/marzer/tomlplusplus/blob/master/CONTRIBUTING.md
[how-to]: https://github.com/marzer/tomlplusplus/blob/master/CONTRIBUTING.md#regenerating-tomlhpp
[README.md]: https://github.com/marzer/tomlplusplus/blob/master/README.md

### CODE REVIEW

File: examples/tipi-example/.tipi/deps
Comment (pysco68) [inline examples/tipi-example/.tipi/deps:3]: Just as a note, this should be changed to `marzer/tomlplusplus` once the changes have been merged, currently pointing to our fork just to have the CI builds pass (otherwise this config would have to be more complex as it would require duplicating what's in `/.tipi/deps`, here's what it would look like:)

```json
{
	"requires": {
		"marzer/tomlplusplus": { 
			"requires": {
				"tipi-deps/Catch2@file://single_include": { "@": "v2.x" },
				"nlohmann/json": { "@": "v3.10.5" }
			}
		}
	}
}
```

**Comment by marzer** [comment]
Hi, thanks for the contribution! tipi seems pretty cool, always happy to support new build systems/package managers/tooling etc.

My only note would be about the addition of `tipi-example`. I'd rather not have another example just for a build system, since it's not really an example of toml++ at all. It's an example of how to use tipi, and belongs there instead.

**Comment by pysco68** [comment]
Sure I can remove that. Just thought it could be valuable to have the CI test that the deps resolution works as expected (we integrate dependencies from source, straight out of your repo)

**Comment by marzer** [comment]
>  Just thought it could be valuable to have the CI test that the deps resolution works

Hmmmn, yeah I suppose that's reasonable. Is there some way to refactor it to not require a separate subdirectory with a superfluous copy+paste of code files? That's the part that's bothering me, really.

**Comment by pysco68** [comment]
I'll try to make that happen 👍  Probably won't happen before Monday though

**Comment by marzer** [comment]
> Probably won't happen before Monday though

No hurry. I'd like to cut a new minor bugfix release at some point soon but there's no fixed timeframe for that.

File: README.md
Comment (marzer) [inline README.md:129]: You should probably also add a section to `docs/pages/main_page.dox` - that's the front page of the [HTML docs](https://marzer.github.io/tomlplusplus/)

File: .tipi/deps
Comment (marzer) [inline .tipi/deps:6]: What are the semantics for tipi dependency resolution? toml++ only requires these for running tests; regular users have no need for these depedencies at all.

**Review by marzer — CHANGES_REQUESTED** [review]

File: README.md
Comment (marzer) [reply README.md:129]: Oh, also, the "Try this" section isn't correct anymore without the dedicated `examples/tipi-example`.

**Review by marzer — CHANGES_REQUESTED** [review]

File: README.md
Comment (marzer) [reply README.md:129]: Oh, also, the "Try this" section isn't correct anymore without the dedicated examples/tipi-example.

**Comment by marzer** [comment]
@pysco68 FYI I've just pushed some CI-related refactors that should bring the CI iteration time down to around 5 mins; consider rebasing your branch on my changes to save yourself some sanity ^_^

**Comment by pysco68** [comment]
@marzer thanks for the heads up 👍

**Comment by pysco68** [comment]
@marzer btw. can you give me a short feedback about my last commit / the refactoring of the dependency use build? Is that okay for you?

I basically generate the example on-the-fly now in the CI script by copying stuff around

**Comment by marzer** [comment]
Oh, yup, I saw it earlier. It looks good to me :)

File: .tipi/deps
Comment (pysco68) [reply .tipi/deps:6]: That's a point we're currently working on. We are trying a bunch of variants to find something that works nicely for many cases (there's some headaches involved when you have repository "features" that collide with each other, and we want to gracefully handle some edge cases). For now the users have to pull the deps anyway but the perf hit is really there only for the first build, after that its cached. Not perfect (yet) but it works

Comment (marzer) [reply .tipi/deps:6]: Ah yup. Sounds reasonable.

Comment (marzer) [reply .tipi/deps:6]: > here's some headaches involved when you have repository "features" that collide with each other

C++ tooling in a nutshell

Comment (pysco68) [reply .tipi/deps:6]: Yeah. But we're pretty confident that we'll turn up with something actually nice. We try hard to not just shift aways the ugly things to our users but rather to find sane defaults that are nice to live with ;)

**Comment by marzer** [comment]
@pysco68 FYI I'm going to do the next release some time tomorrow

**Comment by pysco68** [comment]
@marzer thanks for the heads up. I just merged your latest changes and cleaned up the doc, should be good to go from my side. Sorry for the delay, my week was extremely busy.

**Comment by marzer** [comment]
> Sorry for the delay, my week was extremely busy.

No worries mate, life is complex. Just thought you might like the heads-up if you wanted the tipi docs to make the next release ^_^

**Comment by marzer** [comment]
Ah, in all the excitement to get the next release out I forgot to merge this in after all. Sorry @pysco68.

In any event, it looks good, just needs rebasing against the current master since it touches on some files I've refactored a little since then.

**Review by marzer — APPROVED** [review]

**Comment by pysco68** [comment]
@marzer 🤞 I should be good in a few minutes if I didn't oversee some other dumb GH-action detail

**Comment by pysco68** [comment]
👍 went through both in the PR and in my fork, so I *guess* this shoud be mergeable. Thanks for bothering with me so long ;)

**Comment by pysco68** [comment]
Awesome! If there's any breakage with this just ping me.

**Comment by marzer** [comment]
@pysco68 I'm thinking of just vendoring catch2 and nlohmann/json in the repo directly since I'm tired of dealing with git submodules and other iterop nonsense; what implications does that have for the `.tipi/deps` file if there's no dependencies? Do I jut delete it?

**Comment by marzer** [comment]
@pysco68 The tipi CI jobs have started failing in a way that doesn't make sense to me: https://github.com/marzer/tomlplusplus/actions/runs/3079090281/jobs/4980413514

mind taking a look?

**Comment by pysco68** [comment]
Hey @marzer I just had a look at the log and it's a bug in our latest release, we have a fix that's on its way, it should be released by friday at latest. Sorry for the inconvenience.

I'll be in touch ASAP

**Comment by marzer** [comment]
@pysco68 thanks!

**Comment by marzer** [comment]
@pysco68 Were you able to fix the bug? I ended up disabling the tipi CI altogether.

**Pull request merged**
